Fix hex-float parse

kariya-mitsuru opened the pull request:
Passed new 52 tests.

**before**
```
End ".../wain/spec-test/wasm-testsuite/const.wast":
  total: 767, passed: 707, failed: 60, skipped: 0

Results of 76 files:
  total: 19740, passed: 18662, failed: 882, skipped: 196
```

**after**
```
End ".../wain/spec-test/wasm-testsuite/const.wast":
  total: 767, passed: 759, failed: 8, skipped: 0

Results of 76 files:
  total: 19740, passed: 18714, failed: 830, skipped: 196
```

kariya-mitsuru committed:
Fix hex-float parse

rhysd commented on wain-syntax-text/src/parser.rs:
```diff
@@ -589,56 +589,116 @@ parse_uint_function!(parse_u8_str, u8);
-                } else if c == '.' {
-                    break;
+macro_rules! parse_hex_float_fn {
+    ($name:ident, $float:ty, $uint:ty, $bits:expr) => {
```
`$bits` マクロ引数の代わりに `size_of` const 関数を使っていただけると．マクロ引数の数はできるだけ少なくしておきたいです．

```rust
const BITS: i32 = (std::mem::size_of::<$float>() * 8) as i32;
```

https://doc.rust-lang.org/std/mem/fn.size_of.html

kariya-mitsuru replied:
なんと、こんな関数が！？

rhysd replied:
so benri

rhysd commented on wain-syntax-text/src/parser.rs:
```diff
@@ -589,56 +589,116 @@ parse_uint_function!(parse_u8_str, u8);
+                    offset,
+                );
+            }
+            Ok(sign.apply(f))
```
sign ビットも `m` を構築する時点で立てちゃっても良いかもですね（もちろんこのままでも OK です）

kariya-mitsuru replied:
最初そうしようかなと思ったんですが、無限大判定の時にはまだ符号は無くてよいし、`sign.apply` でも生成コードの実行効率変わらなそうで if も不要だったのでこちらにしました。

rhysd replied:
分かりました，それでまったく問題ないです:+1:

rhysd commented:
お手数ですが，この PR 以前だと fail する入力を教えていただけませんでしょうか？テストケースに追加しておきます．

kariya-mitsuru committed:
use std::mem::size_of instead of argument

rhysd commented on wain-syntax-text/src/parser.rs:
```diff
@@ -589,56 +589,117 @@ parse_uint_function!(parse_u8_str, u8);
-                    break;
+macro_rules! parse_hex_float_fn {
+    ($name:ident, $float:ty, $uint:ty) => {
+        fn $name<'s>(parser: &Parser<'s>, sign: Sign, m_str: &'s str, oexp: Option<(Sign, &'s str)>, offset: usize) -> Result<'s, $float> {
```
もしどこか参考にした資料があれば，URL をコメントで書いておいていただけると助かります．

rhysd commented on wain-syntax-text/src/parser.rs:
```diff
@@ -589,56 +589,117 @@ parse_uint_function!(parse_u8_str, u8);
-                    continue;
-                } else if c == '.' {
-                    break;
+macro_rules! parse_hex_float_fn {
```
もしどこか参考にした資料（できればアルゴリズムの解説）があれば，URL をコメントで書いておいていただけると助かります．アルゴリズムについての情報が何も無い状態で逆算で読むのがしんどいです

kariya-mitsuru replied:
オリジナルなので参考にした資料は無いですね…

rhysd replied:
うーん，そうするとお手数ですが，もう少しコメントをつけておいていただけると助かります（具体例とか付けておいていただけるととてもありがたいです）．コードだけで僕の頭の中にはアルゴリズムが無いので，メンテが厳しいです

kariya-mitsuru replied:
あ、すみません、今まだ日本語で説明を考えている最中でした…
（鳥頭なのでまだちゃんと説明できなくて悩んでいる…）
全部日本語で説明を書き終わったら一応こちらに追記しようと思います。
すみません…

kariya-mitsuru replied:
1. 「指数部」をパースする。（無い場合はゼロ）
結果が i32 の範囲に収まらない場合にはエラーとする。

2. 「仮数部」をパースする。 
round to nearest even をするためには、`SIGNIFICAND_BITS + 1` ビット分のパース結果と、それ以降にゼロ以外の数字があるかの情報が必要なため、`SIGNIFICAND_BITS + 2` ビット以上パースした後は、それ以降の数字にゼロ以外があった場合に LSB を 1 にセットする事で LSB に情報を保存しておく。（この段階ではまだ「仮数部」を正規化していないため、「仮数部」の有効ビット数は `SIGNIFICAND_BITS + 2` ～ `SIGNIFICAND_BITS + 5` の可能性があるが、`SIGNIFICAND_BITS + 1` より下のビットはゼロかゼロ以外かが判別出来れば良いだけなので、ゼロ以外があった場合に LSB を 1 にセットするだけで問題ない）
また、パースした「仮数部」を整数として見た場合に「仮数部*2^指数部」が表現したい数値になるように、「指数部」を調整する。この際、「指数部」が最大値や最小値を超える場合は、どうせこの後で∞やゼロになるだけなので、飽和演算を使用しておく。

3. 「仮数部」を正規化する。
「仮数部」を、LSB を 1 ビット目と数えた際の `SIGNIFICAND_BITS + 5` ビット目が 1 になるように正規化する。
「指数部」もそれに合わせて調整する。なお、ここでも上記と同様の理由で飽和演算を使用しておく。

4. 「指数部」の範囲に合わせたエンコード処理を行う。
    1. 「指数部」が正規化数の範囲の場合
「仮数部」は implicit bit をマスクした後、round to nearest even（※１）を行う。
「指数部」は「仮数部」を整数部 1 桁とみなすための調整（`SIGNIFICAND_BITS - 1` を足す）とゲタ表現のための調整（`fxx::MAX_EXP - 1` を足す※２）を行う。
なお、round to nearest even の処理をした結果桁あふれが発生することがあり、その場合は「仮数部」をゼロにした上で「指数部」に +1 をする必要があるが、桁あふれが発生した時の「仮数部」は「仮数部」部分がゼロでかつ「指数部」のビットに 1 が繰り上がった状態になっているので「仮数部」をマスクせずに加算によって「指数部」との合成をする事で特別な処理を不要にしている。
    2. 「指数部」が非正規化数の範囲の場合
まず、「指数部」が指数の最小値（`fxx::MIN_EXP - 1`）になるように、「仮数部」をシフトする。この時、はみ出てしまう分のビットの中に 1 があった場合には round to nearest even の処理に影響を及ぼす可能性があるため、シフト後の LSB を 1 にセットする。
その後、round to nearest even（※１）の処理を行う。
非正規化数の「指数部」はゼロである。
なお、round to nearest even の処理をした結果桁あふれが発生することがあり、その場合は「仮数部」をゼロにした上で「指数部」を 1 にする必要があるが、桁あふれが発生した時の「仮数部」は「仮数部」部分がゼロでかつ「指数部」のビットに 1 が繰り上がった状態になっているので「仮数部」をマスクせずに使う事で特別な処理を不要にしている。
また、「仮数部」のシフト時に「仮数部」がゼロになる可能性もあるが、その場合でも「指数部」「仮数部」共にゼロとなる事でゼロの表現になるために問題ない。
    3. 「指数部」が正規化数の最大値より大きい場合
∞とする。
    4. 「指数部」が非正規化数の最小値より小さい場合
ゼロとする。

5. エンコード結果を浮動小数点数に変換する。変換された浮動小数点数が∞だった場合は、エラーとする。

6. 符号を処理する。

※１：round to nearest even では、「仮数部」と `0x2f` との論理積の結果がゼロでなければ `0x10` を加える事で処理している。
「仮数部」下位 7 ビットの処理前と処理後の状態は以下の通り。
```
000_0000(切り捨て) -> 000_0000
000_0001(切り捨て) -> 001_0001
000_1111(切り捨て) -> 001_1111
001_0000(切り捨て) -> 000_0000
001_0001(切り上げ) -> 010_0001
001_1111(切り上げ) -> 010_1111
010_0000(切り捨て) -> 010_0000
010_0001(切り捨て) -> 011_0001
010_1111(切り捨て) -> 011_1111
011_0000(切り上げ) -> 100_0000
011_0001(切り上げ) -> 100_0001
011_1111(切り上げ) -> 100_1111
```

※２：Rust の `fxx::MIN_EXP` と `fxx::MAX_EXP` はなぜか本来の値より 1 大きいので、-1 する必要がある。

※３：「仮数部」がゼロの場合は（符号ビット以外は）エンコード結果もゼロなので、3.と4.の処理は不要。

rhysd replied:
ありがとうございます🙏 内容を理解してからコードのほうにもコメントとして追記しておきます．これでメンテのときにも困らなさそうです．

rhysd commented on wain-syntax-text/src/parser.rs:
```diff
@@ -589,56 +589,117 @@ parse_uint_function!(parse_u8_str, u8);
+            const BITS: i32 = (mem::size_of::<$float>() * 8) as i32;
+            const SIGNIFICAND_BITS: i32 = <$float>::MANTISSA_DIGITS as i32;
+            const INFINITY: $uint = (ONE << BITS - SIGNIFICAND_BITS) - 1 << SIGNIFICAND_BITS - 1;
+            const TEMP_SIG_BITS: i32 = SIGNIFICAND_BITS + 5;
```
この + 5 ってなんでしょう？

kariya-mitsuru replied:
round to nearest even を行うために必要なビット数は +2 ビットですが、仮数部が 16 進数で表現されているので一桁増やすと 4 ビットずつ増えていきます。
なので、最低 +2 ビット、最高 +5 ビットということで +5 ビットになります。

rhysd replied:
4 bits ずつ増えていくということは，significand のビット数に桁上りの 4bits + 1bit ということですよね．この 1bit は，下記で質問させていただいた，round to nearest even の「全て1ビット立っているか」をチェックするためのビットに充てているという理解で良いでしょうか？

kariya-mitsuru replied:
「桁上がりの 4bits + 1bit」と言われているのがちょっと分かりませんが、どちらかと言えば「1bit + 4bits」だと思います。
1bit は四捨五入（0捨1入）されるビットで、残りの 4bits は左側の 1bit が 1 で、かつ、更にその上のビットも 1 の時に、切り上げるか切り捨てるかを判断するためのビット群です。

kariya-mitsuru replied:
1. 「指数部」をパースする。（無い場合はゼロ）
結果が i32 の範囲に収まらない場合にはエラーとする。

2. 「仮数部」をパースする。 
round to nearest even をするためには、`SIGNIFICAND_BITS + 1` ビット分のパース結果と、それ以降にゼロ以外の数字があるかの情報が必要なため、`SIGNIFICAND_BITS + 2` ビット以上パースした後は、それ以降の数字にゼロ以外があった場合に LSB を 1 にセットする事で LSB に情報を保存しておく。（この段階ではまだ「仮数部」を正規化していないため、「仮数部」の有効ビット数は `SIGNIFICAND_BITS + 2` ～ `SIGNIFICAND_BITS + 5` の可能性があるが、`SIGNIFICAND_BITS + 1` より下のビットはゼロかゼロ以外かが判別出来れば良いだけなので、ゼロ以外があった場合に LSB を 1 にセットするだけで問題ない）
また、パースした「仮数部」を整数として見た場合に「仮数部*2^指数部」が表現したい数値になるように、「指数部」を調整する。この際、「指数部」が最大値や最小値を超える場合は、どうせこの後で∞やゼロになるだけなので、飽和演算を使用しておく。

3. 「仮数部」を正規化する。
「仮数部」を、LSB を 1 ビット目と数えた際の `SIGNIFICAND_BITS + 5` ビット目が 1 になるように正規化する。
「指数部」もそれに合わせて調整する。なお、ここでも上記と同様の理由で飽和演算を使用しておく。

4. 「指数部」の範囲に合わせたエンコード処理を行う。
    1. 「指数部」が正規化数の範囲の場合
「仮数部」は implicit bit をマスクした後、round to nearest even（※１）を行う。
「指数部」は「仮数部」を整数部 1 桁とみなすための調整（`SIGNIFICAND_BITS - 1` を足す）とゲタ表現のための調整（`fxx::MAX_EXP - 1` を足す※２）を行う。
なお、round to nearest even の処理をした結果桁あふれが発生することがあり、その場合は「仮数部」をゼロにした上で「指数部」に +1 をする必要があるが、桁あふれが発生した時の「仮数部」は「仮数部」部分がゼロでかつ「指数部」のビットに 1 が繰り上がった状態になっているので「仮数部」をマスクせずに加算によって「指数部」との合成をする事で特別な処理を不要にしている。
    2. 「指数部」が非正規化数の範囲の場合
まず、「指数部」が指数の最小値（`uxx::MIN_EXP - 1`）になるように、「仮数部」をシフトする。この時、はみ出てしまう分のビットの中に 1 があった場合には round to nearest even の処理に影響を及ぼす可能性があるため、シフト後の LSB を 1 にセットする。
その後、round to nearest even（※１）の処理を行う。
非正規化数の「指数部」はゼロである。
なお、round to nearest even の処理をした結果桁あふれが発生することがあり、その場合は「仮数部」をゼロにした上で「指数部」を 1 にする必要があるが、桁あふれが発生した時の「仮数部」は「仮数部」部分がゼロでかつ「指数部」のビットに 1 が繰り上がった状態になっているので「仮数部」をマスクせずに使う事で特別な処理を不要にしている。
また、「仮数部」のシフト時に「仮数部」がゼロになる可能性もあるが、その場合でも「指数部」「仮数部」共にゼロとなる事でゼロの表現になるために問題ない。
    3. 「指数部」が正規化数の最大値より大きい場合
∞とする。
    4. 「指数部」が非正規化数の最小値より小さい場合
ゼロとする。

5. エンコード結果を浮動小数点数に変換する。変換された浮動小数点数が∞だった場合は、エラーとする。

6. 符号を処理する。

※１：round to nearest even では、「仮数部」と `0x2f` との論理積の結果がゼロでなければ `0x10` を加える事で処理している。
「仮数部」下位 7 ビットの処理前と処理後の状態は以下の通り。
```
000_0000(切り捨て) -> 000_0000
000_0001(切り捨て) -> 001_0001
000_1111(切り捨て) -> 001_1111
001_0000(切り捨て) -> 000_0000
001_0001(切り上げ) -> 010_0001
001_1111(切り上げ) -> 010_1111
010_0000(切り捨て) -> 010_0000
010_0001(切り捨て) -> 011_0001
010_1111(切り捨て) -> 011_1111
011_0000(切り上げ) -> 100_0000
011_0001(切り上げ) -> 100_0001
011_1111(切り上げ) -> 100_1111
```

※２：Rust の `fxx::MIN_EXP` と `fxx::MAX_EXP` はなぜか本来の値より 1 大きいので、-1 する必要がある。

※３：「仮数部」がゼロの場合は（符号ビット以外は）エンコード結果もゼロなので、3.と4.の処理は不要。

rhysd commented on wain-syntax-text/src/parser.rs:
```diff
@@ -589,56 +589,117 @@ parse_uint_function!(parse_u8_str, u8);
+            const TEMP_SIG_BITS: i32 = SIGNIFICAND_BITS + 5;
+            const TEMP_EXP_BIAS: i32 = <$float>::MAX_EXP - 2 + TEMP_SIG_BITS;
+            const TEMP_MAX_EXP: i32 = <$float>::MAX_EXP - TEMP_SIG_BITS;
+            const TEMP_MIN_EXP: i32 = <$float>::MIN_EXP - TEMP_SIG_BITS;
```
この `TEMP_` プレフィックスはどういう意味でしょう？

kariya-mitsuru replied:
temporary です。

rhysd replied:
すみません，質問の仕方が悪かったのですが，どういう意味で temporary とつけてるのでしょう？

kariya-mitsuru replied:
計算途中（処理中）の significand と exponent なので temporary significand と temporary exponent の意味で temporary と名付けました。
変数名を temp_sig と temp_exp に修正したのですが、これでちょっとは分かりやすくなったでしょうか？

rhysd commented on wain-syntax-text/src/parser.rs:
```diff
@@ -589,56 +589,117 @@ parse_uint_function!(parse_u8_str, u8);
+                            stringify!($float),
+                            format!("exponent out of range '{}{}'", exp_sign.to_string(), exp_str),
+                            offset,
+                        ),
```
このエラーメッセージは `exp_str` のパースに失敗した時に正しくないので，下記のように2通りのエラーメッセージを出してもらえませんでしょうか

```rust
                    Ok(_) => return parser.cannot_parse_num(
                            stringify!($float),
                            format!("exponent value out of range '{}{}'", exp_sign.to_string(), exp_str),
                            offset,
                        ),
                    Err(err) => return parser.cannot_parse_num(
                            stringify!($float),
                            format!("invalid integer for exponent part: '{}{}'", exp_sign.to_string(), exp_str),
                            offset,
                        ),
```

kariya-mitsuru replied:
パースに失敗とはどのような状態を想定しているでしょうか？
lexer 側でチェックしているので、値が u32 に収まらない以外のエラーは発生しないはずなので、2 種類にはならないと思いますが…
（"value" は付けておきます）

rhysd replied:
それであれば，Result を unwrap して，明示的に起き得ないということを示していただけると，単に考慮漏れしてるわけでないことが分かるのでありがたいです

kariya-mitsuru replied:
えぇと、`OK(_)` も `Err(_)` も起き得るのでこのような書き方になっているのですが（`unwrap` はできない）、以下のように同じエラーメッセージを 2 回に分けて書いた方が良いと言うことでしょうか？

```
                    Ok(_) => return parser.cannot_parse_num(
                            stringify!($float),
                            format!("exponent value out of range '{}{}'", exp_sign.to_string(), exp_str),
                            offset,
                        ),
                    Err(_) => return parser.cannot_parse_num(
                            stringify!($float),
                            format!("exponent value out of range '{}{}'", exp_sign.to_string(), exp_str),
exp_str),
                            offset,
                        ),
```

rhysd replied:
分かりました．このあたりのエラーハンドリング（lexer の実装を信用するか）は後で考えようと思います．現状はこのままで大丈夫です

rhysd replied:
（すみません，勘違いしてたことに気付いてコメント削除したのですが入れ違いになってしまいました…）

kariya-mitsuru replied:
同じ処理を 2 回書くのはあんまりなのでコメントを記載しましたが、いかがでしょうか？

kariya-mitsuru replied:
すみません、コードを貼り忘れました…
```
                    _ => {
                        // Ok(_)  -> u32::MAX <= expu < i32::MIN or i32::MAX < expu <= u32::MAX
                        // Err(_) -> exp_str is larger than u32::MAX
                        //           (Lexer guarantees that "invalid digit" cannot occur)
                        return parser.cannot_parse_num(
                            stringify!($float),
                            format!("exponent value out of range '{}{}'", exp_sign.to_string(), exp_str),
                            offset,
                        );
                    }
```

rhysd replied:
ありがとうございます．これで大丈夫です:+1:

rhysd commented on wain-syntax-text/src/parser.rs:
```diff
@@ -589,56 +589,117 @@ parse_uint_function!(parse_u8_str, u8);
+            };
+
+            // parse significand & adjust exponent
+            let mut m: $uint = 0;
```
`m` って mantissa の略でしょうか？寿命がない変数での1文字変数名はなるべく避けてもらえると助かります

kariya-mitsuru replied:
もともと mantissa の m でしたが、mantissa は用語として正しくないらしいので significand を意味する temp_sig ではどうでしょう？

rhysd replied:
はい，だいじょぶです．単に1文字変数だと後から見てわかりにくいかなと思っただけだったのでした

rhysd replied:
もしくは単に significand でも良いかもですが

rhysd commented on wain-syntax-text/src/parser.rs:
```diff
@@ -589,56 +589,117 @@ parse_uint_function!(parse_u8_str, u8);
+                    // there are not enough significant digits
+                    _ if m < ONE << SIGNIFICAND_BITS + 1 => {
+                        // shift significand & append significant digit
+                        m = m << 4 | c.to_digit(16).unwrap() as $uint;
```
ここ `unwrap` して良いのでしたっけ（lexer で弾いてるから OK ということでしょうか？）

kariya-mitsuru replied:
はい、lexer でチェック済みなので失敗しないはずです。

rhysd replied:
了解です

rhysd commented on wain-syntax-text/src/parser.rs:
```diff
@@ -589,56 +589,117 @@ parse_uint_function!(parse_u8_str, u8);
+                    }
+                    // there are enough significant digits
+                    _ => {
+                        // set lsb of significand to 1 if non zoro digit (for "round to nearest even")
```
nit: typo (zoro → zero)

あと僕が

> for "round to nearest even"

を分かっていないからだと思うのですが，次の行の，LSB を桁が 0 でなければ立てる処理がなぜ必要なのか分かりませんでした．教えていただけると助かります

kariya-mitsuru replied:
単なる四捨五入（2進数なので0捨1入？）だと変換時の一時的な仮数部に +1 ビット必要なだけですが、round to nearest even だと +1 ビットでは足りません。
例えば、
```
01.0000000000000000 → 01
01.0000000000000001 → 01
01.0111111111111111 → 01
01.1000000000000000 → 10
01.1000000000000001 → 10
01.1111111111111111 → 10
10.0000000000000000 → 10
10.0000000000000001 → 10
10.0111111111111111 → 10
10.1000000000000000 → 10 ここが単なる0捨1入と違う
10.1000000000000001 → 11
10.1111111111111111 → 11
```
なので、後の判定のために `0.1` の下の桁が全て `0` かそれ以外かをの判定が必要です。
上のコメントに書いたように変換時の一時的な仮数部は最低 +2 ビット、最高 +5 ビットとなるように調整しているので、ここの条件に入るときの最下位ビットはここの例で言う `0.01` から `0.00001` にあたる桁位置になります。
で、全て `0` かそれ以外かが分かればいいので、`0` じゃない時に最下位ビットを `1` にしておけば後の判定に十分ということになります。

rhysd replied:
なるほど，分かりました．ありがとうございます．nearest even で切り捨てか切り上げかの判定が必要になるかどうかの条件を最下位ビットに保持してるのですね．

kariya-mitsuru committed:
Follow Github comments

fix typo (zoro -> zero)
fix error message (exponent -> exponent value)
change var name (m -> temp_sig)

kariya-mitsuru commented:
> お手数ですが，この PR 以前だと fail する入力を教えていただけませんでしょうか？テストケースに追加しておきます．

spec-test で今まで fail していたテストケース 52 個（上にも書きましたが全て const.wast のテストケースです）、手元で抽出はしてみましたがどうしましょうか？
分類としては round to nearest even に関するケース（assert_return 44 ケース）と、値が大きすぎて無限大になるケース（assert_invalid 8 ケース）です。

rhysd commented:
ありがとうございます！助かります．マージ後にテストに入れておきたいので，ここのコメントで良いのでコピペしておいていただけると．

kariya-mitsuru committed:
Follow Github comments

add comment
change var name (exp -> temp_exp)

kariya-mitsuru commented:
52 テストケースです。
<details><code><pre>
(assert_malformed
  (module quote "(func (f32.const 0x1p128) drop)")
  "constant out of range"
)
(assert_malformed
  (module quote "(func (f32.const -0x1p128) drop)")
  "constant out of range"
)
(assert_malformed
  (module quote "(func (f32.const 0x1.ffffffp127) drop)")
  "constant out of range"
)
(assert_malformed
  (module quote "(func (f32.const -0x1.ffffffp127) drop)")
  "constant out of range"
)

(assert_malformed
  (module quote "(func (f64.const 0x1p1024) drop)")
  "constant out of range"
)
(assert_malformed
  (module quote "(func (f64.const -0x1p1024) drop)")
  "constant out of range"
)
(assert_malformed
  (module quote "(func (f64.const 0x1.fffffffffffff8p1023) drop)")
  "constant out of range"
)
(assert_malformed
  (module quote "(func (f64.const -0x1.fffffffffffff8p1023) drop)")
  "constant out of range"
)

;; Rounding behaviour

;; f32, small exponent
(module (func (export "f") (result f32) (f32.const +0x1.00000100000000001p-50)))
(assert_return (invoke "f") (f32.const +0x1.000002p-50))
(module (func (export "f") (result f32) (f32.const -0x1.00000100000000001p-50)))
(assert_return (invoke "f") (f32.const -0x1.000002p-50))
(module (func (export "f") (result f32) (f32.const +0x1.000001fffffffffffp-50)))
(assert_return (invoke "f") (f32.const +0x1.000002p-50))
(module (func (export "f") (result f32) (f32.const -0x1.000001fffffffffffp-50)))
(assert_return (invoke "f") (f32.const -0x1.000002p-50))
(module (func (export "f") (result f32) (f32.const +0x1.00000500000000001p-50)))
(assert_return (invoke "f") (f32.const +0x1.000006p-50))
(module (func (export "f") (result f32) (f32.const -0x1.00000500000000001p-50)))
(assert_return (invoke "f") (f32.const -0x1.000006p-50))

(module (func (export "f") (result f32) (f32.const +0x4000.004000001p-64)))
(assert_return (invoke "f") (f32.const +0x1.000002p-50))
(module (func (export "f") (result f32) (f32.const -0x4000.004000001p-64)))
(assert_return (invoke "f") (f32.const -0x1.000002p-50))
(module (func (export "f") (result f32) (f32.const +0x4000.014000001p-64)))
(assert_return (invoke "f") (f32.const +0x1.000006p-50))
(module (func (export "f") (result f32) (f32.const -0x4000.014000001p-64)))
(assert_return (invoke "f") (f32.const -0x1.000006p-50))

;; f32, large exponent
(module (func (export "f") (result f32) (f32.const +0x1.00000100000000001p+50)))
(assert_return (invoke "f") (f32.const +0x1.000002p+50))
(module (func (export "f") (result f32) (f32.const -0x1.00000100000000001p+50)))
(assert_return (invoke "f") (f32.const -0x1.000002p+50))
(module (func (export "f") (result f32) (f32.const +0x1.000001fffffffffffp+50)))
(assert_return (invoke "f") (f32.const +0x1.000002p+50))
(module (func (export "f") (result f32) (f32.const -0x1.000001fffffffffffp+50)))
(assert_return (invoke "f") (f32.const -0x1.000002p+50))
(module (func (export "f") (result f32) (f32.const +0x1.00000500000000001p+50)))
(assert_return (invoke "f") (f32.const +0x1.000006p+50))
(module (func (export "f") (result f32) (f32.const -0x1.00000500000000001p+50)))
(assert_return (invoke "f") (f32.const -0x1.000006p+50))

(module (func (export "f") (result f32) (f32.const +0x4000004000001)))
(assert_return (invoke "f") (f32.const +0x1.000002p+50))
(module (func (export "f") (result f32) (f32.const -0x4000004000001)))
(assert_return (invoke "f") (f32.const -0x1.000002p+50))


;; f32, subnormal
(module (func (export "f") (result f32) (f32.const +0x0.00000100000000001p-126)))
(assert_return (invoke "f") (f32.const +0x0.000002p-126))
(module (func (export "f") (result f32) (f32.const -0x0.00000100000000001p-126)))
(assert_return (invoke "f") (f32.const -0x0.000002p-126))
(module (func (export "f") (result f32) (f32.const +0x0.000002fffffffffffp-126)))
(assert_return (invoke "f") (f32.const +0x0.000002p-126))
(module (func (export "f") (result f32) (f32.const -0x0.000002fffffffffffp-126)))
(assert_return (invoke "f") (f32.const -0x0.000002p-126))
(module (func (export "f") (result f32) (f32.const +0x0.00000500000000001p-126)))
(assert_return (invoke "f") (f32.const +0x0.000006p-126))
(module (func (export "f") (result f32) (f32.const -0x0.00000500000000001p-126)))
(assert_return (invoke "f") (f32.const -0x0.000006p-126))

;; f64, small exponent
(module (func (export "f") (result f64) (f64.const +0x1.000000000000080000000001p-600)))
(assert_return (invoke "f") (f64.const +0x1.0000000000001p-600))
(module (func (export "f") (result f64) (f64.const -0x1.000000000000080000000001p-600)))
(assert_return (invoke "f") (f64.const -0x1.0000000000001p-600))
(module (func (export "f") (result f64) (f64.const +0x1.000000000000280000000001p-600)))
(assert_return (invoke "f") (f64.const +0x1.0000000000003p-600))
(module (func (export "f") (result f64) (f64.const -0x1.000000000000280000000001p-600)))
(assert_return (invoke "f") (f64.const -0x1.0000000000003p-600))

(module (func (export "f") (result f64) (f64.const +0x8000000.000000400000000001p-627)))
(assert_return (invoke "f") (f64.const +0x1.0000000000001p-600))
(module (func (export "f") (result f64) (f64.const -0x8000000.000000400000000001p-627)))
(assert_return (invoke "f") (f64.const -0x1.0000000000001p-600))
(module (func (export "f") (result f64) (f64.const +0x8000000.000001400000000001p-627)))
(assert_return (invoke "f") (f64.const +0x1.0000000000003p-600))
(module (func (export "f") (result f64) (f64.const -0x8000000.000001400000000001p-627)))
(assert_return (invoke "f") (f64.const -0x1.0000000000003p-600))


;; f64, large exponent
(module (func (export "f") (result f64) (f64.const +0x1.000000000000080000000001p+600)))
(assert_return (invoke "f") (f64.const +0x1.0000000000001p+600))
(module (func (export "f") (result f64) (f64.const -0x1.000000000000080000000001p+600)))
(assert_return (invoke "f") (f64.const -0x1.0000000000001p+600))
(module (func (export "f") (result f64) (f64.const +0x1.000000000000280000000001p+600)))
(assert_return (invoke "f") (f64.const +0x1.0000000000003p+600))
(module (func (export "f") (result f64) (f64.const -0x1.000000000000280000000001p+600)))
(assert_return (invoke "f") (f64.const -0x1.0000000000003p+600))

(module (func (export "f") (result f64) (f64.const +0x2000000000000100000000001)))
(assert_return (invoke "f") (f64.const +0x1.0000000000001p+97))
(module (func (export "f") (result f64) (f64.const -0x2000000000000100000000001)))
(assert_return (invoke "f") (f64.const -0x1.0000000000001p+97))
(module (func (export "f") (result f64) (f64.const +0x20000000000001fffffffffff)))
(assert_return (invoke "f") (f64.const +0x1.0000000000001p+97))
(module (func (export "f") (result f64) (f64.const -0x20000000000001fffffffffff)))
(assert_return (invoke "f") (f64.const -0x1.0000000000001p+97))
(module (func (export "f") (result f64) (f64.const +0x2000000000000500000000001)))
(assert_return (invoke "f") (f64.const +0x1.0000000000003p+97))
(module (func (export "f") (result f64) (f64.const -0x2000000000000500000000001)))
(assert_return (invoke "f") (f64.const -0x1.0000000000003p+97))

;; f64, subnormal
(module (func (export "f") (result f64) (f64.const +0x1.000000000000280000000001p-1022)))
(assert_return (invoke "f") (f64.const +0x1.0000000000003p-1022))
(module (func (export "f") (result f64) (f64.const -0x1.000000000000280000000001p-1022)))
(assert_return (invoke "f") (f64.const -0x1.0000000000003p-1022))
</pre></code></details>

rhysd commented on wain-syntax-text/src/parser.rs:
```diff
@@ -589,56 +589,122 @@ parse_uint_function!(parse_u8_str, u8);
+            const INFINITY: $uint = (ONE << BITS - SIGNIFICAND_BITS) - 1 << SIGNIFICAND_BITS - 1;
+            const TEMP_SIG_BITS: i32 = SIGNIFICAND_BITS + 5;
+            const TEMP_EXP_BIAS: i32 = <$float>::MAX_EXP - 2 + TEMP_SIG_BITS;
+            const TEMP_MAX_EXP: i32 = <$float>::MAX_EXP - TEMP_SIG_BITS;
```
指数部にバイアスをかける処理がまだ理解できてないのですが，`TEMP_EXP_BIAS` の `- 2 + TEMP_SIG_BITS` と `TEMP_MAX_EXP` の `- TEMP_SIG_BITS` はどういう計算を行っているのでしょう？（例えば f32 のバイアスは `0b11111111 - f32::MAX_EXP = 127` ですよね）

kariya-mitsuru replied:
`temp_sig` に格納されている significand は、encode 時には下から `TEMP_SIG_BITS` ビット目が必ず `1` になるように normalize 処理しています。
また、`temp_exp` に格納されている exponent は、`significand` を整数としたときの exponent ですが（exponent がなかったときが 0 なので）、最終的に encode された exponent では significand の整数部が 1 ビットになった時の値です。
そこで `TEMP_SIG_BITS - 1` を足す必要があります。
また exponent のゲタは exponent の最大値と同じなのですが、`f32::MAX_EXP` とかがなぜか 1 大きいので `MAX_EXP - 1` を足す必要があります。（`f32::MIN_EXP` も 1 大きい）
以上から、`TEMP_SIG_BITS - 1 + MAX_EXP - 1` で `MAX_EXP - 2 + TEMP_SIG_BITS` になります。

kariya-mitsuru replied:
`TEMP_MAX_EXP` も同様で、normalize 処理後の `temp_exp` は encode するための exponent よりも `TEMP_SIG_BITS - 1` だけ小さいので、`temp_exp` を比較して処理を場合分けするためにはあらかじめ `TEMP_SIG_BITS - 1` だけ引いておく必要があるのですが、`MAX_EXP` や `MIN_EXP` がなぜかそもそも 1 大きいので `TEMP_SIG_BITS` だけ引いています。

rhysd replied:
説明ありがとうございます．自分の理解を確認しておきたいのですが，

> 最終的に encode された exponent では significand の整数部が 1 ビットになった時の値です

これはつまり，エンコード前の `significand` は整数値で，例えば `0b11010101` が入っていますが，binary-32 にエンコードするときは `0b1.1010101` に桁を合わせないといけないため，その桁をずらす分の exp のビット数が `TEMP_SIG_BITS - 1` という理解で良いでしょうか？

kariya-mitsuru replied:
はい、そうです。

rhysd replied:
ありがとうございます．多分一通り把握できたと思います．もしまたわからなくなったら，すみませんがツイッターとかで質問させてください．
